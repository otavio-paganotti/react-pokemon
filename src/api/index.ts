/**
 * Exporting the Axios Instance
 */
export * from './api';

/**
 * Exporting all modules
 */
export * as Pokemon from './services/rest/pokemon';
