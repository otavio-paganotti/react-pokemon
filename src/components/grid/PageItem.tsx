import clsx from 'clsx';

const PageItem = ({
  item,
  page,
  itemLabel,
  current,
}: {
  item: number;
  itemLabel: number | string;
  page?: (page: number) => void;
  current?: boolean;
}) => {
  return (
    <div
      className={clsx('page-item', current ? 'current' : '')}
      onClick={() => page && page(item)}
    >
      {itemLabel}
    </div>
  );
};

export default PageItem;
