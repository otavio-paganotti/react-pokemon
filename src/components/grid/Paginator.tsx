import clsx from "clsx";

const PageItem = ({ item, page, itemLabel, current }: {
  item: number;
  itemLabel: number | string;
  page?: ((page: number) => void);
  current?: boolean;
}) => {
  return (
    <div className={clsx('page-item', current ? 'current' : '')} onClick={() => page && page(item)}>
      {itemLabel}
    </div>
  )
};

const Paginator = ({
  pagination,
  page
}: {
  pagination: {
    firstPage: number;
    lastPage: number;
    nextPages: number[];
    prevPages: number[];
    nextPage: number;
    prevPage: number;
    page: number;
  },
  page: ((page: number) => void)
}) => {
  return (
    <div className="flex justify-center mb-4">
      <div className="mb-4">
        <nav className="flex gap-4" aria-label="Pagination">
          <PageItem page={(item) => pagination.page > 1 && page(item)} item={pagination.firstPage} itemLabel="<<" />
          <PageItem page={(item) => pagination.page > 0 && page(item)} item={pagination.prevPage} itemLabel="<" />
          {pagination.prevPages.length > 0 && pagination.prevPages.map((item) => (
            <PageItem item={item} itemLabel={item} page={page} />
          ))}
          <PageItem item={pagination.page} itemLabel={pagination.page} current={true} />
          {pagination.nextPages.length > 0 && pagination.nextPages.map((item) => (
            <PageItem item={item} itemLabel={item} page={page} />
          ))}
          <PageItem page={(item) => pagination.page < pagination.lastPage && page(item)} item={pagination.nextPage} itemLabel=">" />
          <PageItem page={(item) => pagination.page < pagination.lastPage && page(item)} item={pagination.lastPage} itemLabel=">>" />
        </nav>
      </div>
    </div>
  )
}

export default Paginator;