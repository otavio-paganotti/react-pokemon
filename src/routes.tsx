import type { RouteObject } from 'react-router';
import MainLayout from '@/layouts/Default';
import Home from '@/pages/index/Index'
import Pokemon from '@/pages/pokemon/Index'

const routes: RouteObject[] = [
  {
    path: '*',
    element: <MainLayout />,
    children: [
      {
        index: true,
        element: <Home />,
      },
      {
        index: false,
        path: ':id',
        element: <Pokemon />
      }
    ],
  },
];

export default routes;