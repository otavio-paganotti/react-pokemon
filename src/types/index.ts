export type Nullable<T = unknown> = T | null;

export interface PokemonItem {
  name: string;
  url: string;
  id: number;
  content: Pokemon;
}

export interface Response<T> {
  count: number;
  next: Nullable<string>;
  previous: Nullable<string>;
  results: T[];
  pagination: {
    firstPage: number;
    lastPage: number;
    nextPage: number;
    prevPage: number;
    nextPages: number[];
    prevPages: number[];
    page: number;
  };
}

export interface Sprites {
  back_default: string;
}

export interface Stat {
  base_stat: number;
  effort: number;
  stat: Species;
}

export interface Type {
  slot: number;
  type: Species;
}

export interface VersionGroupDetail {
  level_learned_at: number;
  move_learn_method: Species;
  version_group: Species;
}

export interface Move {
  move: Species;
  version_group_details: VersionGroupDetail[];
}

export interface VersionDetail {
  rarity: number;
  version: Species;
}

export interface Species {
  name: string;
  url: string;
}

export interface HeldItem {
  item: Species;
  version_details: VersionDetail[];
}

export interface GameIndex {
  game_index: number;
}

export interface Ability {
  ability: Species;
  is_hidden: boolean;
  slot: number;
}

export interface Pokemon {
  abilities: Ability[];
  base_experience: number;
  forms: Species[];
  game_indices: GameIndex[];
  height: number;
  held_items: HeldItem[];
  id: number;
  is_default: boolean;
  location_area_encounters: string;
  moves: Move[];
  name: string;
  order: number;
  past_types: unknown[];
  species: Species;
  sprites: Sprites;
  stats: Stat[];
  types: Type[];
  weight: number;
}

export interface IPagination {
  firstPage: number;
  lastPage: number;
  nextPages: number[];
  prevPages: number[];
  nextPage: number;
  prevPage: number;
  page: number;
}
