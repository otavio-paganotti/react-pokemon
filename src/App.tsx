import routes from '@/routes';
import { useRoutes } from 'react-router-dom';
import '@/assets/css/main.scss';

const App = () => {
  const content = useRoutes(routes);

  return content;
};

export default App;
