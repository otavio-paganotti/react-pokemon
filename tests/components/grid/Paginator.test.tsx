import React from 'react';
import renderer from 'react-test-renderer';
import Paginator from '../../../src/components/grid/Paginator';
import PageItem from '../../../src/components/grid/PageItem';
import { describe, beforeEach, expect, test, vi } from 'vitest';
import { IPagination } from '../../../src/types';

vi.mock('../../../src/components/grid/PageItem');

const toJson = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
};

const paginationMock = {
  firstPage: 1,
  lastPage: 1,
  nextPage: 1,
  prevPage: 1,
  nextPages: [],
  prevPages: [],
  page: 1,
};

vi.mock('react-router-dom');

const page = (page: number) => {
  vi.fn();
};

const onMounted = async (mock: IPagination) => {
  let component;

  component = renderer.create(
    <Paginator
      page={page}
      pagination={mock}
    />
  );

  return toJson(component);
};

describe('Paginator.tsx', () => {
  let tree;

  beforeEach(async () => {
    vi.clearAllMocks();

    tree = await onMounted(paginationMock);
  });

  test('should render correctly the Paginator component and take a snapshot', () => {
    expect(tree).toMatchSnapshot();
  });

  test('should have 5 PageItem child element must be called', async () => {
    expect(PageItem).toHaveBeenCalled();
    
    expect(PageItem).toHaveBeenCalledTimes(5);
  });

  test('shoud have 9 PageItem when we are on first page', async () => {
    vi.clearAllMocks();

    tree = await onMounted({
      firstPage: 1,
      lastPage: 10,
      nextPage: 2,
      prevPage: 0,
      nextPages: [2, 3, 4, 5],
      prevPages: [],
      page: 1,
    });

    expect(PageItem).toHaveBeenCalled();

    expect(PageItem).toHaveBeenCalledTimes(9);

    const pageItemProps = [
      {
        page,
        item: 1,
        itemLabel: '<<',
      },
      {
        page,
        item: 0,
        itemLabel: '<',
      },
      {
        item: 1,
        itemLabel: 1,
        current: true,
      },
      {
        page,
        item: 2,
        itemLabel: 2,
      },
      {
        page,
        item: 3,
        itemLabel: 3,
      },
      {
        page,
        item: 4,
        itemLabel: 4,
      },
      {
        page,
        item: 5,
        itemLabel: 5,
      },
      {
        page,
        item: 2,
        itemLabel: '>',
      },
      {
        page,
        item: 10,
        itemLabel: '>>',
      },
    ];

    pageItemProps.forEach((props, index) => {
      expect(PageItem).toHaveBeenNthCalledWith(index + 1, props, {});
    });
  });

  test('shoud have 9 PageItem, when we are on last page', async () => {
    vi.clearAllMocks();

    tree = await onMounted({
      firstPage: 1,
      lastPage: 10,
      nextPage: 0,
      prevPage: 9,
      nextPages: [],
      prevPages: [6, 7, 8, 9],
      page: 10,
    });

    expect(PageItem).toHaveBeenCalled();

    expect(PageItem).toHaveBeenCalledTimes(9);

    const pageItemProps = [
      {
        page,
        item: 1,
        itemLabel: '<<',
      },
      {
        page,
        item: 9,
        itemLabel: '<',
      },
      {
        page,
        item: 6,
        itemLabel: 6,
      },
      {
        page,
        item: 7,
        itemLabel: 7,
      },
      {
        page,
        item: 8,
        itemLabel: 8,
      },
      {
        page,
        item: 9,
        itemLabel: 9,
      },
      {
        item: 10,
        itemLabel: 10,
        current: true,
      },
      {
        page,
        item: 0,
        itemLabel: '>',
      },
      {
        page,
        item: 10,
        itemLabel: '>>',
      },
    ];

    pageItemProps.forEach((props, index) => {
      expect(PageItem).toHaveBeenNthCalledWith(index + 1, props, {});
    });
  });

  test('shoud have 9 PageItem, when we are on middle page', async () => {
    vi.clearAllMocks();

    tree = await onMounted({
      firstPage: 1,
      lastPage: 10,
      nextPage: 5,
      prevPage: 4,
      nextPages: [6, 7],
      prevPages: [3, 4],
      page: 5,
    });

    expect(PageItem).toHaveBeenCalled();

    expect(PageItem).toHaveBeenCalledTimes(9);

    const pageItemProps = [
      {
        page,
        item: 1,
        itemLabel: '<<',
      },
      {
        page,
        item: 4,
        itemLabel: '<',
      },
      {
        page,
        item: 3,
        itemLabel: 3,
      },
      {
        page,
        item: 4,
        itemLabel: 4,
      },
      {
        item: 5,
        itemLabel: 5,
        current: true,
      },
      {
        page,
        item: 6,
        itemLabel: 6,
      },
      {
        page,
        item: 7,
        itemLabel: 7,
      },
      {
        page,
        item: 5,
        itemLabel: '>',
      },
      {
        page,
        item: 10,
        itemLabel: '>>',
      },
    ];

    pageItemProps.forEach((props, index) => {
      expect(PageItem).toHaveBeenNthCalledWith(index + 1, props, {});
    });
  });
});
