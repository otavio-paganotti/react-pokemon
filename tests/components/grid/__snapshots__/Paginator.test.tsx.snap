// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Paginator.tsx > should render correctly the Paginator component and take a snapshot 1`] = `
<div
  className="flex justify-center mb-4"
>
  <div
    className="mb-4"
  >
    <nav
      aria-label="Pagination"
      className="flex md:gap-4"
    />
  </div>
</div>
`;
